### PR TITLE
Add is_polynomial function

### PR DESCRIPTION
**Pull request by rikardn**
Implemented using a visitor. Instead of calling `free_symbols` as sympy does it operates in two modes. One where only symbols are considered. This makes early returns possible and will make it faster. All unit tests from sympy have been incorporated.

### CODE REVIEW

**Comment by rikardn** [comment]
Something is strange with the code coverage. The figures are green when I check the details.

**Comment by lkeegan** [comment]
> Something is strange with the code coverage. The figures are green when I check the details.

looks like some error occurred with the upload to codecov on this build:
https://github.com/symengine/symengine/runs/2155451207?check_suite_focus=true#step:3:11744
maybe restart all jobs to try again? the pr looks good to me by the way!

**Comment by rikardn** [comment]
What happened was that I first had a reduction in coverage. I fixed that by adding tests and force pushed. For some reason the new codecov-value hasn't been picked up by github. I restarted all github actions tests now. Not sure if codecov will be updated by this. It might not matter for this particular PR, but if this happens often it might become annoying.

**Comment by lkeegan** [comment]
@rikardn I think the new coverage value is being updated on github, but it's still showing a reduction in coverage (of the whole project, not of the code in your pr) because one of the two builds that upload coverage data is not succeeding in uploading the results to codecov:
build error message: https://github.com/symengine/symengine/runs/2155451207?check_suite_focus=true#step:3:11744
possibly related: https://github.com/codecov/codecov-action/issues/68

**Comment by rikardn** [comment]
Thanks @lkeegan for the detective work. Let's not have the flakiness of codecov come between this PR and its reviewer :wink:

File: symengine/test_visitors.h [hunk 1]
```diff
@@ -213,6 +213,40 @@ class RealVisitor : public BaseVisitor<RealVisitor>
...
+    };
+    void bvisit(const Relational &x)
+    {
+        is_polynomial_ = false;
+    };
+
+    bool apply(const Basic &b);
+};
+
+bool is_polynomial(const Basic &b, const set_basic &variables = {});
```
Comment (isuruf) [inline symengine/test_visitors.h:249]: I don't understand what `variables` is doing. Can you explain with a comment?

Comment (rikardn) [reply symengine/test_visitors.h:259]: I added a documentation comment. The `variables` is analogous to `syms` in sympy. See also https://github.com/sympy/sympy/issues/21088 for an explanation.

**Review by isuruf — APPROVED** [review]

**Comment by rikardn** [comment]
Thanks @lkeegan and @isuruf for reviewing

**Pull request merged**